### ISSUE TITLE
[RFC] Should we have a canary branch

### ISSUE THREAD

**Issue by rossriley** [bolt/bolt#2385]
Ok, so I'm fairly well along with the test suite for Bolt, looking at around 60% code coverage so far.

There's a few changes stacking up though that are going to result in a fairly big refactor that whilst not breaking backwards compatibility would definitely be candidates for next minor version. 

The problem I can see ahead is that if we want to keep master branch relatively stable merging in larger refactors is going to be a big job and I'm aware that other contributors would probably much rather develop new features against the test suite and associated refactors.

So the suggestion is that we have a common unstable canary branch that we continually rebase against master and that all new features targeting next minor release get PR'd into. Thoughts?

**Comment by GawainLynch** [comment]
Totally in favour. This is one reason why I was pushing for GitFlow over an agile approach like GitHib Flow.

**Comment by bobdenotter** [comment]
I was actually thinking about that, the other day.. I'm going to work on an update to Bootstrap 3.3, but it will break a few minor things, until I find and fix them.. So, :+1: 

Oh, and Happy new year!! ^_^

**Comment by rossriley** [comment]
Happy new year too. :-)

So am i right in that the best way to manage this  is to create the branch but treat it the same as master in that no changes are ever pushed to it directly and no merges are made using this branch. 

Instead PRs are made and merged from our own feature branches and then once decided as stable we create a PR using the same feature branch on master. 

That way the canary branch is effectively a way to test many various feature branches in conjunction, but we don't run into a situation where we need to do big merges since the feature branches will have proven themselves pretty stable.

As a nice side effect we should have a situation where it's easy to scan the changes to master by looking purely at which feature branches have been merged in.

**Comment by GawainLynch** [comment]
Happy New Year while we're at it :-)

Yes, I agree @rossriley, the carany branch idea as proposed sound really good to me.

**Comment by rixbeck** [comment]
Happy New Year! :+1:

**Comment by bobdenotter** [comment]
So, now that we agree on this, how exactly should we word this? Basically, we'll have three branches:

- master (aka "2.x stable")
- canary ("2.x bleeding edge")
- legacy ("1.x stable)"

All _new_ features should be merged into Canary. Bugfixes and minor tweaks go directly into Master, right? 

But, how will we manage to keep these in sync? Let's say i'm working on "feature X", and Ross on "Feature Y". My first batch of commits is merged into canary, and Ross's is too. Mine has a bunch of bugs, and Ross's has a few less bugs. Ross fixes them, and they are merged into canary. How are we then going to merge Ross's feature into Master, without also merging my half-baked work?

**Comment by GawainLynch** [comment]
> But, how will we manage to keep these in sync?

GitFlow would be the obvious retort here, but not really useful.

The obvious way to deal with this actually gets a bit complicated for new contributors, but pretty straight forward for core devs...  Basically we could/would assume `Canary` branch can be wiped and resynced with `master` at any point.  In the meantime the person merging into `Canary` just gets to deal with conflicts, but what is PR'd to master is the *individual* feature branch itself.

Having said that, I want to cry at the idea of continuously managing and communicating that to/with most contributors. I think we're venturing into trying to combine agile and waterfall, and they not really compatible methodologies.

**Comment by rossriley** [comment]
Agree with @GawainLynch canary needs to track master, but i'm not seeing how this is vastly more complex to what we have done previously.

Bugs / Fixes need to happen on the feature branch. Once the feature branch is proved stable via testing for a few weeks on canary, then it becomes a candidate for merge onto master, once the branch is merged onto master it gets deleted and any other further fixes / features would need to start with a new feature.

For cases where a few contributors want to collaborate on a feature then an RFC could lead to the creation of a a feature branch on the bolt repo itself rather than an individual fork, and bug fixes / simple fixes could bypass the canary branch altogether if deemed suitable for a minor point release.

**Comment by GawainLynch** [comment]
Canaries are GO! :bird:
